Front end diffs

mattheu opened the pull request:
This is a V1 of the front end history/diffs functionality. This is built using react.

I'd love for someone with some react knowlege to give this once over and let me know if i'm `doing_it_wrong`!

Todo
- [x] Use the API to fetch the data, don't just use localize script.
- [x] Maybe the UI should be expandable/collapsable? Collapsed by default? Fetch data on request? Feedback wanted.
- [x] Handle more than 10 revisions. Some kind of pagination or load more functionality.

mattheu commented:
This is what it looks like right now.

![image](https://cloud.githubusercontent.com/assets/494927/16762402/3942c75c-481c-11e6-99de-7185fae3187b.png)

![image](https://cloud.githubusercontent.com/assets/494927/16762423/4e3badea-481c-11e6-935a-d8c50544d937.png)

rmccue commented on assets/src/scripts/inc/history.js:
```diff
@@ -0,0 +1,282 @@
+	},
+
+	render: function() {
+		return ( <div className="PageHistory_Diff" dangerouslySetInnerHTML={ { __html: this.getDiffHTML() } } /> );
```
Rather than setting HTML, since you're building this dynamically, it'd be better to use `<div>{ this.getDiffHTML() }</div>` and return elements from `getDiffHTML()`

rmccue commented on assets/src/scripts/inc/history.js:
```diff
@@ -0,0 +1,282 @@
+/**
+ * History List.
+ */
+var PageHistoryList = React.createClass({
```
Generally, these should be one-class-per-file.

rmccue commented on assets/src/scripts/inc/history.js:
```diff
@@ -0,0 +1,282 @@
+			<div className="PageHistory_List_Container">
+				<h4 className="PageHistory_List_Title">Page History</h4>
+				<ul className="PageHistory_List">
+					{ this.state.revisions.map( function( revision ) {
```
ES6 syntax for the callback: `.map( revision => { ... })`

rmccue commented on assets/src/scripts/inc/history.js:
```diff
@@ -0,0 +1,282 @@
+
+						return (
+							<PageHistoryListItem key={ revision.id } id={ revision.id } action={ revision.action } author={ revision.author } date={ revision.date } content={ revision.content } active={ revision.active } actions={ actions } />
+						);
```
Parens aren't necessary for a single item. `return <PageHistoryListItem ...`

rmccue replied:
You can also use `{...revision}` to "spread" the object as parameters, since you're doing one-to-one mapping here.

rmccue commented on assets/src/scripts/inc/history.js:
```diff
@@ -0,0 +1,282 @@
+		}
+
+		return (
+			<li className={ classNames.join( ' ' ) } onClick={ this.handleSelectRevision }>
```
Maybe use an anonymous for the click handler? `e => this.props.actions.selectRevision( this.props )`

Generally, action callbacks are called `on...` rather than being under a subkey as well.

rmccue commented on assets/src/scripts/inc/history.js:
```diff
@@ -0,0 +1,282 @@
+				<PageHistoryDiff diff_a={ this.state.diff.a } diff_b={ this.state.diff.b } />
+				<PageHistoryList revisions={ this.state.revisions } actions={ actions } />
+			</div>
+		);
```
Parens unnecessary here.

rmccue commented on assets/src/scripts/inc/history.js:
```diff
@@ -0,0 +1,282 @@
+		// Defer until actually set.
+		window.setTimeout( function() {
+			this.toggleContainerClass();
+		}.bind( this ) );
```
ES6 closures automatically bind, so can be `window.setTimeout( () => this.toggleContainerClass() );`

rmccue commented on assets/src/scripts/inc/history.js:
```diff
@@ -0,0 +1,282 @@
+			return;
+		}
+
+		console.log( this.state.diff );
```
🙅

mattheu commented:
Anybody able to give this another review? Pretty much rewritten following @rmccue's scathing comments.

mattheu commented:
Only outstanding thing - i'm not so keen on the load more button style. Just seems very heavy.

rmccue commented on .babelrc:
```diff
@@ -0,0 +1 @@
+ { "presets": ["react"] }
```
This should probably be in your `package.json` instead, avoids needing the separate file.

rmccue commented on assets/src/scripts/components/PageHistory/PageHistory.jsx:
```diff
@@ -0,0 +1,181 @@
+
+	    super( props );
+
+		this.state = {
```
Indentation is off (and lots of excess newlines here.

rmccue commented on assets/src/scripts/components/PageHistory/PageHistory.jsx:
```diff
@@ -0,0 +1,181 @@
+import React from 'react';
+import PageHistoryDiff from './PageHistoryDiff.jsx';
+import PageHistoryList from './PageHistoryList.jsx';
```
Can drop the `.jsx` here if you add it to the supported extensions in your webpack config. Also, prefer `.js` over `.jsx` generally.

rmccue commented on assets/src/scripts/components/PageHistory/PageHistory.jsx:
```diff
@@ -0,0 +1,181 @@
+		};
+
+		return (
+			<div>
```
Single line for this:
```js
return <div>
    <Page ... />
</div>;
```

rmccue commented on assets/src/scripts/components/PageHistory/PageHistory.jsx:
```diff
@@ -0,0 +1,181 @@
+
+		var id     = window.HMHandbookPageHistory.post_id;
+		var base   = window.HMHandbookPageHistory.api_base;
+		var nonce  = window.HMHandbookPageHistory.api_nonce;
```
Globals should be avoided in favour of either props or context

rmccue replied:
(Although the API bits are probably OK to use it, the ID isn't)

mattheu replied:
Moved all these global settings to a settings file (along with the strings).

Post_ID still comes from the global, but is passed into the component when created.

rmccue commented on assets/src/scripts/components/PageHistory/PageHistory.jsx:
```diff
@@ -0,0 +1,181 @@
+				_revision.active = true;
+			} else {
+				_revision.active = false;
+			}
```
Could condense this to `_revision.active = revision && revision.id === _revision.id`

rmccue commented on assets/src/scripts/components/PageHistory/PageHistory.jsx:
```diff
@@ -0,0 +1,181 @@
+		// Defer until actually set.
+		window.setTimeout( () => {
+			this.toggleContainerClass();
+		} );
```
The closing one shouldn't have extra whitespace:
```js
window.setTimeout( () => {
    // ...
});
```

Also, generally reconsider the need to use setTimeout.

mattheu replied:
It seems I can remove all of these. I had originally run into some issues with `setState` that were fixed by using this - but I must have been doing something wrong.

rmccue commented on assets/src/scripts/components/PageHistory/PageHistory.jsx:
```diff
@@ -0,0 +1,181 @@
+		if  ( null === this.state.diff.a || null === this.state.diff.b ) {
+			container[0].classList.remove( 'article-showing-diff' );
+		} else {
+			container[0].classList.add( 'article-showing-diff' );
```
Is this outside of React? Shouldn't use it for React components.

mattheu replied:
Yes. This is the container element as I want to toggle a class on this depending on the state of the react component.

I think the neatest thing here would be store the container as a property on the component.

rmccue commented on assets/src/scripts/components/PageHistory/PageHistory.jsx:
```diff
@@ -0,0 +1,181 @@
+
+PageHistory.defaultProps = {
+	revisions: []
+}
```
Missing semicolon for consistency. Also, crazy amounts of whitespace in this file!

rmccue commented on assets/src/scripts/components/PageHistory/PageHistoryDiff.jsx:
```diff
@@ -0,0 +1,54 @@
+export default class PageHistoryDiff extends React.Component {
+
+	render() {
+		return (
```
`return <div>`

rmccue commented on assets/src/scripts/components/PageHistory/PageHistoryDiff.jsx:
```diff
@@ -0,0 +1,54 @@
+PageHistoryDiff.defaultProps = {
+	diff_a: '',
+	diff_b: '',
+}
```
Semicolon

rmccue commented on assets/src/scripts/components/PageHistory/PageHistoryList.jsx:
```diff
@@ -0,0 +1,106 @@
+
+		this.state = {
+			expanded:  false,
+		};
```
Indentation

rmccue commented on assets/src/scripts/components/PageHistory/PageHistoryList.jsx:
```diff
@@ -0,0 +1,106 @@
+			onfetchRevisions: this.props.actions.onFetchRevisions,
+		};
+
+		containerClasses = [	'PageHistory_List_Container' ];
```
Whitespace

rmccue commented on assets/src/scripts/components/PageHistory/PageHistoryList.jsx:
```diff
@@ -0,0 +1,106 @@
+
+						return <PageHistoryListItem key={ revision.id } { ...revision } actions={ actions } />;
+
+					} ) }
```
`})}`

rmccue commented on assets/src/scripts/components/PageHistory/PageHistoryList.jsx:
```diff
@@ -0,0 +1,106 @@
+
+					} ) }
+				</ul>
+				<button onClick={ () => { actions.onfetchRevisions() } } className={ loadMorebuttonClasses.join( ' ' ) } disabled={ this.props.hasMore ? null : 'disabled' }>
```
The callback doesn't need the extra `{}`, should be `={ () => actions.onFetchRevisions() }`

rmccue replied:
Because these are DOM nodes, I think you can set `disabled` to a bool instead

rmccue commented on assets/src/scripts/components/PageHistory/PageHistoryList.jsx:
```diff
@@ -0,0 +1,106 @@
+		window.setTimeout( () => {
+			this.maybeLoadRevisions();
+			this.maybeClearDiff();
+		});
```
This seems like an unnecessary setTimeout call?

rmccue commented on assets/src/scripts/components/PageHistory/PageHistoryList.jsx:
```diff
@@ -0,0 +1,106 @@
+	revisions: [],
+	loading:   false,
+	hasMore:   true,
+}
```
Semicolon

rmccue commented on assets/src/scripts/components/PageHistory/PageHistoryListItem.jsx:
```diff
@@ -0,0 +1,31 @@
+		}
+
+		return (
+			<li className={ classNames.join( ' ' ) } onClick={ e => { this.props.actions.onSelectRevision( this.props ) } }>
```
`return <li>`

rmccue replied:
Also, excess `{}` on the callback

rmccue commented on gulpfile.js:
```diff
@@ -1,18 +1,18 @@
 		outputStyle: 'compressed'
 	},
-	webpack: require( './webpack.config.js' ),
+	webpack: require( './.webpack-config.js' ),
```
`webpack.config.js` is the built-in name that `webpack` recognises automatically, so better to use that over a dotfile

mattheu commented:
Thanks for your review @rmccue!

This is probably ready to go.

mattheu commented:
One last thing I've added is a permissions callback to ensure this endoint can't be used to leak unpublished data.  @sambulance you happy for this to be merged?

johnbillion commented:
If we're going to expose revisions on the front end, I really like the idea of revision notes: https://en-gb.wordpress.org/plugins/revision-notes/

johnbillion commented on inc/page-history.php:
```diff
@@ -0,0 +1,125 @@
+			'loadMore'  => __( 'Load more revisions', 'hm-handbook' ),
+		],
+		'post_id'   => get_the_ID(),
+		'api_base'  => home_url( '/wp-json/hm-handbook/v1/' ),
```
This should use `rest_url()`.

mattheu commented:
I like the idea of revision notes. I think it should be tackled separately though. Keen to get this out and get some feedback.

I fixed the rest URL. And changed the logic a little bit to load this based on post type support so you can more easily remove in child themes if you don't want it.

sambulance commented on assets/src/styles/components/_History.scss:
```diff
@@ -0,0 +1,99 @@
+}
+
+.History_Diff-Removed {
+	background: lighten( $hm-primary, 35% );;
```
Extra semicolon here

sambulance commented on assets/src/styles/components/_History.scss:
```diff
@@ -0,0 +1,99 @@
+
+.History_Diff-Added {
+	background: #daf2d6;
```
Should we make this a variable? Looks like it's used in a few places and can then be customised on child themes if need be

mattheu replied:
Yeah that makes sense.

sambulance commented on assets/src/styles/components/_PageHistory.scss:
```diff
@@ -0,0 +1,238 @@
+}
+
+.PageHistory_Diff-Removed {
+	background: lighten( $hm-primary, 35% );;
```
Double semicolon here too.. these styles are a duplicate of the file above, can we use one style for both instances?

sambulance commented on assets/src/styles/theme.scss:
```diff
@@ -13,3 +13,24 @@ $hm-images: "./../../../vendor/hm-pattern-library/assets/images";
 @import "components/article";
+@import "components/PageHistory";
+
+// .NavAccordion_Item-Active > .NavAccordion_Anchor {
```
Remove this if it isn't needed?

sambulance commented on inc/page-history.php:
```diff
@@ -0,0 +1,161 @@
+	add_action( 'rest_api_init',      __NAMESPACE__ . '\\setup_api', 20 );
+
+	foreach ( [ 'post', 'page' ] as $post_type ) {
+		add_post_type_support( 'post', 'hm-handbook-page-history' );
```
Should `post` be `$post_type` here?

sambulance commented on inc/page-history.php:
```diff
@@ -0,0 +1,161 @@
+ */
+function get_supported_post_types() {
+	return array_filter( get_post_types(), function( $post_type ) {
+		return post_type_supports( $post_type, 'hm-handbook-page-history' );
```
Oh wait misread `post_type_supports` 👍
